### PR TITLE
gcc12, gfortran12, gnat12: init at 12.1.0

### PR DESCRIPTION
**Pull request by trofi**
A copy of gcc11 with minor patch changes:
- follow .c/.cc rename upstream
- made patches unconditional

Changes: https://www.gnu.org/software/gcc/gcc-12/changes.html

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Synthetica9** [comment]
Diff between 11 and 12:

```patch
diff '--color=auto' pkgs/development/compilers/gcc/11/default.nix pkgs/development/compilers/gcc/12/default.nix
54,55c54,55
< let majorVersion = "11";
<     version = "${majorVersion}.2.0";
---
> let majorVersion = "12";
>     version = "${majorVersion}.1.0";
61,71c61,67
<       ++ optional noSysDirs ../no-sys-dirs.patch
<       ++ optional (noSysDirs && hostPlatform.isRiscV) ../no-sys-dirs-riscv.patch
<       /* ++ optional (hostPlatform != buildPlatform) (fetchpatch { # XXX: Refine when this should be applied
<         url = "https://git.busybox.net/buildroot/plain/package/gcc/${version}/0900-remove-selftests.patch?id=11271540bfe6adafbc133caf6b5b902a816f5f02";
<         sha256 = ""; # TODO: uncomment and check hash when available.
<       }) */
<       ++ optional langAda ../gnat-cflags-11.patch
<       ++ optional langFortran ../gfortran-driving.patch
<       ++ optional (targetPlatform.libc == "musl" && targetPlatform.isPower) ../ppc-musl.patch
<
<       ++ optional (stdenv.isDarwin && stdenv.isAarch64) (fetchpatch {
---
>       ++ optional noSysDirs ../gcc-12-no-sys-dirs.patch
>       ++ optional noSysDirs ../no-sys-dirs-riscv.patch
>       ++ [
>         ../gnat-cflags-11.patch
>         ../gcc-12-gfortran-driving.patch
>         ../ppc-musl.patch
>       ] ++ optional (stdenv.isDarwin && stdenv.isAarch64) (fetchpatch {
94c90
<     sha256 = "sha256-0I7cU2tUw3KhAQ/2YZ3SdMDxYDqkkhK6IPeqLNo2+os=";
---
>     sha256 = "sha256-Yv1jSInzHAK2SvLEaPBktHrRynhBHEWr5qxLX43RnHs=";
```

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -12510,6 +12510,7 @@ with pkgs;
   gcc9Stdenv = overrideCC gccStdenv buildPackages.gcc9;
   gcc10Stdenv = overrideCC gccStdenv buildPackages.gcc10;
   gcc11Stdenv = overrideCC gccStdenv buildPackages.gcc11;
+  gcc12Stdenv = overrideCC gccStdenv buildPackages.gcc11;
```
Comment (yl3dy) [inline pkgs/top-level/all-packages.nix:12513]: Isn't is a typo (`gcc11` instead of `gcc12`)?

Comment (trofi) [reply pkgs/top-level/all-packages.nix:12513]: Good catch! Updated to `gcc12`.

**Review by yl3dy — APPROVED** [review]
LGTM, have successfully built deal.II master with gcc12 with all tests passing.

**Review by vcunat — APPROVED** [review]
Looks good and works well for me.

**Pull request merged**